dbms/src/Storages/TrivialBuffer.cpp

```diff
@@ -0,0 +1,554 @@
+}
+
+
+StoragePtr TrivialBuffer::create(const std::string & name_, NamesAndTypesListPtr columns_,
+    const NamesAndTypesList & materialized_columns_,
+    const NamesAndTypesList & alias_columns_,
+    const ColumnDefaults & column_defaults_,
+    Context & context_, const size_t num_blocks_to_deduplicate_,
+    const Thresholds & min_thresholds_, const Thresholds & max_thresholds_,
+    const String & destination_database_, const String & destination_table_)
+{
+    return make_shared(
+        name_, columns_, materialized_columns_, alias_columns_, column_defaults_,
+        context_, num_blocks_to_deduplicate_, min_thresholds_, max_thresholds_,
+        destination_database_, destination_table_);
+}
+
+
+TrivialBuffer::TrivialBuffer(const std::string & name_, NamesAndTypesListPtr columns_,
+    const NamesAndTypesList & materialized_columns_,
+    const NamesAndTypesList & alias_columns_,
+    const ColumnDefaults & column_defaults_,
+    Context & context_, const size_t num_blocks_to_deduplicate_,
+    const Thresholds & min_thresholds_, const Thresholds & max_thresholds_,
+    const String & destination_database_, const String & destination_table_)
+    : IStorage{materialized_columns_, alias_columns_, column_defaults_},
+    name(name_), columns(columns_), context(context_),
+    num_blocks_to_deduplicate(num_blocks_to_deduplicate_),
+    current_hashes(std::make_unique<DeduplicationBuffer>()),
+    previous_hashes(std::make_unique<DeduplicationBuffer>()),
+    min_thresholds(min_thresholds_), max_thresholds(max_thresholds_),
+    destination_database(destination_database_), destination_table(destination_table_),
+    no_destination(destination_database.empty() && destination_table.empty()),
+    log(&Logger::get("TrivialBuffer (" + name + ")")),
+    flush_thread(&TrivialBuffer::flushThread, this)
+{
+}
+
+class TrivialBufferBlockInputStream : public IProfilingBlockInputStream
+{
+public:
+    TrivialBufferBlockInputStream(const Names & column_names_, TrivialBuffer & buffer_)
+        : column_names(column_names_.begin(), column_names_.end()), buffer(buffer_) {}
+
+    String getName() const { return "TrivialBuffer"; }
+
+    String getID() const
+    {
+        std::stringstream res;
+        res << "TrivialBuffer(" << &buffer;
+
+        for (const auto & name : column_names)
+            res << ", " << name;
+
+        res << ")";
+        return res.str();
+    }
+
+protected:
+    Block readImpl()
+    {
+        Block res;
+
+        if (has_been_read)
+            return res;
+        has_been_read = true;
+
+        std::lock_guard<std::mutex> lock(buffer.mutex);
+
+        for (auto & block : buffer.data)
+        {
+            if (!block.rows())
+                continue;
+
+            for (const auto & name : column_names)
+            {
+                auto & col = block.getByName(name);
+                res.insert(ColumnWithTypeAndName(col.column->clone(), col.type, name));
```

Comment:
Wrong code in case of more than one block in buffer.
You should implement as in StorageMemory.

Cloning is not necessary.